File: proof_kernel/zero_validation.py
"""Cross-check generated zero assets against reference zero assets."""

from __future__ import annotations

import gzip
import json
import os
from decimal import Decimal, InvalidOperation
from pathlib import Path
from typing import Any, Dict, Iterator, List, Optional, Tuple

from .data_assets import ROOT, hash_file, relpath, utc_now, write_json

DEFAULT_ZERO_REFERENCE_TOLERANCE = "5e-9"
OVERKILL_VALIDATION_COUNT = 60000
OVERKILL_REQUIRED_STORED_DPS = 100


def _asset_path(asset: Dict[str, Any], root: Path = ROOT) -> Optional[Path]:
    source_path = asset.get("source_path")
    if not source_path:
        return None
    path = Path(str(source_path))
    return path if path.is_absolute() else root / path


def _validation_path(asset: Dict[str, Any], root: Path = ROOT) -> Path:
    path = _asset_path(asset, root)
    if path is None:
        asset_id = str(asset.get("asset_id") or "unknown_zero_asset")
        return root / "data" / "zeros" / "nontrivial" / f"{asset_id}.validation.json"
    name = path.name
    for suffix in (".jsonl.gz", ".txt.gz", ".jsonl", ".txt", ".gz"):
        if name.endswith(suffix):
            return path.with_name(name[: -len(suffix)] + ".validation.json")
    return path.with_suffix(".validation.json")


def _iter_lines(path: Path) -> Iterator[str]:
    if path.suffix == ".gz":
        with gzip.open(path, "rt", encoding="utf-8", errors="replace") as handle:
            yield from handle
        return
    with open(path, "r", encoding="utf-8", errors="replace") as handle:
        yield from handle


def _numeric_token(line: str) -> Optional[str]:
    stripped = line.strip()
    if not stripped or stripped.startswith("#"):
        return None
    parts = stripped.split()
    if not parts:
        return None
    token = parts[-1].strip().strip('"').strip("'")
    try:
        Decimal(token)
    except (InvalidOperation, ValueError):
        return None
    return token


def _iter_zero_values(path: Path) -> Iterator[Decimal]:
    for line in _iter_lines(path):
        token = _numeric_token(line)
        if token is not None:
            yield Decimal(token)


def _iter_zero_tokens(path: Path) -> Iterator[Tuple[str, Decimal]]:
    for line in _iter_lines(path):
        token = _numeric_token(line)
        if token is not None:
            yield token, Decimal(token)


def reference_declared_decimals(reference_asset: Optional[Dict[str, Any]]) -> Optional[int]:
    if not isinstance(reference_asset, dict):
        return None
    for key in ("stored_dps", "usable_dps", "declared_decimals"):
        value = reference_asset.get(key)
        try:
            parsed = int(value)
        except (TypeError, ValueError):
            continue
        if parsed > 0:
            return parsed
    return None


def tolerance_for_reference_asset(reference_asset: Optional[Dict[str, Any]]) -> str:
    declared = reference_declared_decimals(reference_asset)
    if declared is None:
        return DEFAULT_ZERO_REFERENCE_TOLERANCE
    return f"5e-{declared}"


def _percentile(sorted_values: List[Decimal], pct: Decimal) -> Decimal:
    if not sorted_values:
        return Decimal(0)
    if len(sorted_values) == 1:
        return sorted_values[0]
    rank = (Decimal(len(sorted_values) - 1) * pct) / Decimal(100)
    lower = int(rank.to_integral_value(rounding="ROUND_FLOOR"))
    upper = int(rank.to_integral_value(rounding="ROUND_CEILING"))
    if lower == upper:
        return sorted_values[lower]
    fraction = rank - Decimal(lower)
    return sorted_values[lower] + (sorted_values[upper] - sorted_values[lower]) * fraction


def _not_available(
    asset: Dict[str, Any],
    reference_asset: Optional[Dict[str, Any]],
    tolerance: str,
    reason: str,
    root: Path,
    max_count: Optional[int] = None,
) -> Dict[str, Any]:
    asset_path = _asset_path(asset, root)
    reference_path = _asset_path(reference_asset, root) if isinstance(reference_asset, dict) else None
    payload = {
        "asset_id": asset.get("asset_id"),
        "generated_asset_path": relpath(asset_path, root) if asset_path else None,
        "reference_asset_id": reference_asset.get("asset_id") if isinstance(reference_asset, dict) else None,
        "reference_asset_path": relpath(reference_path, root) if reference_path else None,
        "validated_count": 0,
        "reference_declared_decimals": reference_declared_decimals(reference_asset),
        "generated_stored_dps": asset.get("stored_dps"),
        "tolerance": str(tolerance),
        "max_deviation": None,
        "p95_deviation": None,
        "failed_indices": [],
        "failed_details": [],
        "status": "NOT_AVAILABLE",
        "valid_for_overkill": False,
        "valid_for_authoritative": False,
        "validation_limit": int(max_count) if max_count is not None else None,
        "created_at": utc_now(),
        "reason": reason,
    }
    write_json(_validation_path(asset, root), payload)
    return payload


def _cached_validation_is_current(
    cached: Dict[str, Any],
    asset: Dict[str, Any],
    reference_asset: Dict[str, Any],
    tolerance: str,
    max_count: Optional[int] = None,
) -> bool:
    if max_count is not None and int(cached.get("validation_limit") or 0) != int(max_count):
        return False
    return (
        cached.get("asset_id") == asset.get("asset_id")
        and cached.get("reference_asset_id") == reference_asset.get("asset_id")
        and str(cached.get("tolerance")) == str(tolerance)
        and cached.get("asset_hash") == asset.get("hash")
        and cached.get("reference_asset_hash") == reference_asset.get("hash")
        and cached.get("status") in {"PASS", "FAIL"}
    )


def validate_zero_asset_against_reference(
    asset: Dict[str, Any],
    reference_asset: Optional[Dict[str, Any]],
    tolerance: str | Decimal = DEFAULT_ZERO_REFERENCE_TOLERANCE,
    root: Path = ROOT,
    max_count: Optional[int] = None,
) -> Dict[str, Any]:
    """Compare overlapping zero indices and persist a validation artifact."""

    tolerance_text = str(tolerance)
    try:
        tol = Decimal(tolerance_text)
    except InvalidOperation:
        tol = Decimal(DEFAULT_ZERO_REFERENCE_TOLERANCE)
        tolerance_text = DEFAULT_ZERO_REFERENCE_TOLERANCE

    asset_path = _asset_path(asset, root)
    if asset_path is None or not asset_path.exists():
        return _not_available(asset, reference_asset, tolerance_text, "asset file is unavailable", root, max_count)
    if not reference_asset:
        return _not_available(asset, reference_asset, tolerance_text, "reference zero asset is unavailable", root, max_count)
    reference_path = _asset_path(reference_asset, root)
    if reference_path is None or not reference_path.exists():
        return _not_available(asset, reference_asset, tolerance_text, "reference zero asset is unavailable", root, max_count)

    validation_path = _validation_path(asset, root)
    if validation_path.exists():
        try:
            cached = json.loads(validation_path.read_text(encoding="utf-8"))
            if _cached_validation_is_current(cached, asset, reference_asset, tolerance_text, max_count):
                return cached
        except Exception:
            pass

    deviations: List[Decimal] = []
    failed_indices: List[int] = []
    failed_details: List[Dict[str, Any]] = []
    max_deviation = Decimal(0)
    count = 0

    for candidate_pair, reference_pair in zip(_iter_zero_tokens(asset_path), _iter_zero_tokens(reference_path)):
        if max_count is not None and count >= int(max_count):
            break
        candidate_text, candidate = candidate_pair
        reference_text, reference = reference_pair
        count += 1
        deviation = abs(candidate - reference)
        deviations.append(deviation)
        if deviation > max_deviation:
            max_deviation = deviation
        if deviation > tol:
            failed_indices.append(count)
            if len(failed_details) < 20:
                failed_details.append({
                    "index": int(count),
                    "generated_value": candidate_text,
                    "reference_value": reference_text,
                    "deviation": str(deviation),
                    "tolerance": tolerance_text,
                })

    if count == 0:
        return _not_available(asset, reference_asset, tolerance_text, "no overlapping zero indices", root, max_count)

    sorted_deviations = sorted(deviations)
    p95_deviation = _percentile(sorted_deviations, Decimal(95))
    status = "PASS" if not failed_indices else "FAIL"
    generated_stored_dps = asset.get("stored_dps")
    try:
        generated_stored_dps_int = int(generated_stored_dps)
    except (TypeError, ValueError):
        generated_stored_dps_int = 0
    valid_for_overkill = (
        status == "PASS"
        and int(count) >= OVERKILL_VALIDATION_COUNT
        and generated_stored_dps_int >= OVERKILL_REQUIRED_STORED_DPS
    )
    payload = {
        "asset_id": asset.get("asset_id"),
        "reference_asset_id": reference_asset.get("asset_id"),
        "asset_path": relpath(asset_path, root),
        "generated_asset_path": relpath(asset_path, root),
        "reference_asset_path": relpath(reference_path, root),
        "asset_hash": asset.get("hash") or hash_file(asset_path),
        "reference_asset_hash": reference_asset.get("hash") or hash_file(reference_path),
        "validated_count": int(count),
        "reference_declared_decimals": reference_declared_decimals(reference_asset),
        "generated_stored_dps": generated_stored_dps,
        "tolerance": tolerance_text,
        "max_deviation": str(max_deviation),
        "p95_deviation": str(p95_deviation),
        "failed_indices": failed_indices,
        "failed_details": failed_details,
        "status": status,
        "valid_for_overkill": valid_for_overkill,
        "valid_for_authoritative": status == "PASS",
        "validation_limit": int(max_count) if max_count is not None else None,
        "created_at": utc_now(),
        "reason": None if status == "PASS" else f"{len(failed_indices)} zero ordinates exceeded tolerance",
    }
    write_json(validation_path, payload)
    return payload


def load_zero_validation_artifacts(root: Path = ROOT) -> List[Dict[str, Any]]:
    out: List[Dict[str, Any]] = []
    validation_root = root / "data" / "zeros" / "nontrivial"
    if not validation_root.exists():
        return out
    for path in sorted(validation_root.glob("*.validation.json")):
        try:
            parsed = json.loads(path.read_text(encoding="utf-8"))
            if isinstance(parsed, dict):
                parsed.setdefault("path", relpath(path, root))
                out.append(parsed)
        except Exception:
            continue
    return out


if __name__ == "__main__":
    print(json.dumps({"validations": load_zero_validation_artifacts()}, indent=2, sort_keys=True))

File: lib/zero-validation.ts
import fs from "fs";
import path from "path";
import zlib from "zlib";
import type { DataAsset } from "./data-assets";

export const DEFAULT_ZERO_REFERENCE_TOLERANCE = "5e-9";
const OVERKILL_VALIDATION_COUNT = 60_000;
const OVERKILL_REQUIRED_STORED_DPS = 100;

export interface ZeroValidationArtifact {
    asset_id?: string;
    reference_asset_id?: string | null;
    asset_path?: string;
    generated_asset_path?: string | null;
    reference_asset_path?: string;
    asset_hash?: unknown;
    reference_asset_hash?: unknown;
    validated_count: number;
    reference_declared_decimals?: number | null;
    generated_stored_dps?: number | null;
    tolerance: string;
    max_deviation: string | null;
    p95_deviation: string | null;
    failed_indices: number[];
    failed_details?: Array<{
        index: number;
        generated_value: string;
        reference_value: string;
        deviation: string;
        tolerance: string;
    }>;
    status: "PASS" | "FAIL" | "NOT_AVAILABLE";
    valid_for_overkill?: boolean;
    valid_for_authoritative: boolean;
    validation_limit?: number | null;
    created_at?: string;
    reason?: string | null;
    path?: string;
}

const repoPath = (candidate?: string | null) => {
    if (!candidate) return null;
    return path.isAbsolute(candidate) ? candidate : path.join(process.cwd(), candidate);
};

const repoRelative = (candidate: string) => {
    try {
        return path.relative(process.cwd(), candidate).replace(/\\/g, "/");
    } catch {
        return candidate.replace(/\\/g, "/");
    }
};

const assetPath = (asset?: DataAsset | Partial<DataAsset> | null) =>
    repoPath(typeof asset?.source_path === "string" ? asset.source_path : null);

const validationPath = (asset: DataAsset | Partial<DataAsset>) => {
    const source = assetPath(asset);
    if (!source) {
        return path.join(
            process.cwd(),
            "data",
            "zeros",
            "nontrivial",
            `${String(asset.asset_id ?? "unknown_zero_asset")}.validation.json`,
        );
    }
    const base = source.replace(/(\.jsonl|\.txt|\.gz)+$/i, "");
    return `${base}.validation.json`;
};

const readFileText = (filePath: string) => {
    const raw = fs.readFileSync(filePath);
    return filePath.endsWith(".gz") ? zlib.gunzipSync(raw).toString("utf8") : raw.toString("utf8");
};

const numericToken = (line: string) => {
    const trimmed = line.trim();
    if (!trimmed || trimmed.startsWith("#")) return null;
    const token = trimmed.split(/\s+/).pop()?.replace(/^["']|["']$/g, "");
    if (!token) return null;
    return Number.isFinite(Number(token)) ? token : null;
};

const zeroTokens = (filePath: string, maxCount?: number | null) => {
    const out: Array<{ token: string; value: number }> = [];
    for (const line of readFileText(filePath).split(/\r?\n/)) {
        if (maxCount !== undefined && maxCount !== null && out.length >= maxCount) break;
        const token = numericToken(line);
        if (token !== null) out.push({ token, value: Number(token) });
    }
    return out;
};

export const referenceDeclaredDecimals = (referenceAsset?: DataAsset | Partial<DataAsset> | null) => {
    for (const key of ["stored_dps", "usable_dps", "declared_decimals"] as const) {
        const parsed = Number(referenceAsset?.[key]);
        if (Number.isFinite(parsed) && parsed > 0) return Math.trunc(parsed);
    }
    return null;
};

export const toleranceForReferenceAsset = (referenceAsset?: DataAsset | Partial<DataAsset> | null) => {
    const declared = referenceDeclaredDecimals(referenceAsset);
    return declared ? `5e-${declared}` : DEFAULT_ZERO_REFERENCE_TOLERANCE;
};

const percentile = (values: number[], pct: number) => {
    if (values.length === 0) return 0;
    if (values.length === 1) return values[0];
    const rank = ((values.length - 1) * pct) / 100;
    const lower = Math.floor(rank);
    const upper = Math.ceil(rank);
    if (lower === upper) return values[lower];
    const fraction = rank - lower;
    return values[lower] + (values[upper] - values[lower]) * fraction;
};

const writeValidation = (asset: DataAsset | Partial<DataAsset>, payload: ZeroValidationArtifact) => {
    const outPath = validationPath(asset);
    fs.mkdirSync(path.dirname(outPath), { recursive: true });
    fs.writeFileSync(outPath, `${JSON.stringify(payload, null, 2)}\n`);
    return { ...payload, path: repoRelative(outPath) };
};

const notAvailable = (
    asset: DataAsset | Partial<DataAsset>,
    referenceAsset: DataAsset | Partial<DataAsset> | null | undefined,
    tolerance: string,
    reason: string,
    maxCount?: number | null,
): ZeroValidationArtifact =>
    writeValidation(asset, {
        asset_id: asset.asset_id,
        generated_asset_path: assetPath(asset) ? repoRelative(assetPath(asset) as string) : null,
        reference_asset_id: referenceAsset?.asset_id ?? null,
        reference_asset_path: assetPath(referenceAsset) ? repoRelative(assetPath(referenceAsset) as string) : undefined,
        validated_count: 0,
        reference_declared_decimals: referenceDeclaredDecimals(referenceAsset),
        generated_stored_dps: typeof asset.stored_dps === "number" ? asset.stored_dps : null,
        tolerance,
        max_deviation: null,
        p95_deviation: null,
        failed_indices: [],
        failed_details: [],
        status: "NOT_AVAILABLE",
        valid_for_overkill: false,
        valid_for_authoritative: false,
        validation_limit: maxCount ?? null,
        created_at: new Date().toISOString(),
        reason,
    });

const readJson = <T>(filePath: string): T | null => {
    try {
        if (!fs.existsSync(filePath)) return null;
        return JSON.parse(fs.readFileSync(filePath, "utf8")) as T;
    } catch {
        return null;
    }
};

const cachedIsCurrent = (
    cached: ZeroValidationArtifact | null,
    asset: DataAsset | Partial<DataAsset>,
    referenceAsset: DataAsset | Partial<DataAsset>,
    tolerance: string,
    maxCount?: number | null,
) => {
    if (!cached) return false;
    if (maxCount !== undefined && maxCount !== null && Number(cached.validation_limit ?? 0) !== maxCount) return false;
    return cached.asset_id === asset.asset_id &&
    cached.reference_asset_id === referenceAsset.asset_id &&
    cached.tolerance === tolerance &&
    cached.asset_hash === asset.hash &&
    cached.reference_asset_hash === referenceAsset.hash &&
    (cached.status === "PASS" || cached.status === "FAIL");
};

export const validateZeroAssetAgainstReference = (
    asset: DataAsset | Partial<DataAsset>,
    referenceAsset: DataAsset | Partial<DataAsset> | null | undefined,
    tolerance = DEFAULT_ZERO_REFERENCE_TOLERANCE,
    maxCount?: number | null,
): ZeroValidationArtifact => {
    const source = assetPath(asset);
    if (!source || !fs.existsSync(source)) {
        return notAvailable(asset, referenceAsset, tolerance, "asset file is unavailable", maxCount);
    }
    const reference = assetPath(referenceAsset);
    if (!referenceAsset || !reference || !fs.existsSync(reference)) {
        return notAvailable(asset, referenceAsset, tolerance, "reference zero asset is unavailable", maxCount);
    }

    const outPath = validationPath(asset);
    const cached = readJson<ZeroValidationArtifact>(outPath);
    if (cachedIsCurrent(cached, asset, referenceAsset, tolerance, maxCount) && cached) {
        return { ...cached, path: repoRelative(outPath) };
    }

    const assetZeros = zeroTokens(source, maxCount);
    const referenceZeros = zeroTokens(reference, maxCount);
    const count = Math.min(assetZeros.length, referenceZeros.length);
    if (count <= 0) {
        return notAvailable(asset, referenceAsset, tolerance, "no overlapping zero indices", maxCount);
    }

    const tol = Number(tolerance);
    const deviations: number[] = [];
    const failed: number[] = [];
    const failedDetails: NonNullable<ZeroValidationArtifact["failed_details"]> = [];
    let maxDeviation = 0;
    for (let idx = 0; idx < count; idx += 1) {
        const deviation = Math.abs(assetZeros[idx].value - referenceZeros[idx].value);
        deviations.push(deviation);
        if (deviation > maxDeviation) maxDeviation = deviation;
        if (deviation > tol) {
            failed.push(idx + 1);
            if (failedDetails.length < 20) {
                failedDetails.push({
                    index: idx + 1,
                    generated_value: assetZeros[idx].token,
                    reference_value: referenceZeros[idx].token,
                    deviation: String(deviation),
                    tolerance,
                });
            }
        }
    }
    deviations.sort((a, b) => a - b);
    const status = failed.length === 0 ? "PASS" : "FAIL";
    const generatedStoredDps = Number(asset.stored_dps ?? 0);
    return writeValidation(asset, {
        asset_id: asset.asset_id,
        reference_asset_id: referenceAsset.asset_id,
        asset_path: repoRelative(source),
        generated_asset_path: repoRelative(source),
        reference_asset_path: repoRelative(reference),
        asset_hash: asset.hash,
        reference_asset_hash: referenceAsset.hash,
        validated_count: count,
        reference_declared_decimals: referenceDeclaredDecimals(referenceAsset),
        generated_stored_dps: Number.isFinite(generatedStoredDps) ? generatedStoredDps : null,
        tolerance,
        max_deviation: String(maxDeviation),
        p95_deviation: String(percentile(deviations, 95)),
        failed_indices: failed,
        failed_details: failedDetails,
        status,
        valid_for_overkill:
            status === "PASS" &&
            count >= OVERKILL_VALIDATION_COUNT &&
            Number.isFinite(generatedStoredDps) &&
            generatedStoredDps >= OVERKILL_REQUIRED_STORED_DPS,
        valid_for_authoritative: status === "PASS",
        validation_limit: maxCount ?? null,
        created_at: new Date().toISOString(),
        reason: status === "PASS" ? null : `${failed.length} zero ordinates exceeded tolerance`,
    });
};

export const readZeroValidationArtifacts = (): ZeroValidationArtifact[] => {
    const dir = path.join(process.cwd(), "data", "zeros", "nontrivial");
    if (!fs.existsSync(dir)) return [];
    const artifacts: ZeroValidationArtifact[] = [];
    for (const name of fs.readdirSync(dir).filter((item) => item.endsWith(".validation.json")).sort()) {
        const filePath = path.join(dir, name);
        const parsed = readJson<ZeroValidationArtifact>(filePath);
        if (parsed) artifacts.push({ ...parsed, path: repoRelative(filePath) });
    }
    return artifacts;
};
